webrender/src/resource_cache.rs

```diff
@@ -546,16 +553,53 @@ impl ResourceCache {
                 .update(image_key, Arc::clone(blob), dirty_rect);
         }
 
+        // Each cache entry stores its own copy of the image's dirty rect. This allows them to be
+        // updated independently. If we are tiling we need to scan the whole cache otherwise there
+        // is a finite number of keys.
+        //
+        // It is important to never assume an empty dirty rect implies a full reupload here,
+        // although we are able to do so elsewhere. We must store the descriptor's full rect
+        // instead. This is because there are update sequences which could cause us to forget the
+        // correct dirty regions if we cleared the dirty rect when we received None, e.g.:
+        //
+        //      1) Update with no dirty rect. We want to reupload everything.
+        //      2) Update with dirty rect B. We still want to reupload everything, not just B.
+        //      3) Perform the upload some time later.
+        if tiling.is_some() {
+            for (request, entry) in self.cached_images.iter_mut() {
```

Comment:
Let's leave a warning here that this might need optimization to avoid the O(N * M) complexity (where N is the number of updates, and M is the number of cache entries).